dbms/src/Flash/Mpp/MinTSOScheduler.h

```diff
@@ -43,31 +43,48 @@ class MinTSOScheduler : private boost::noncopyable
 
     /// delete this to-be cancelled/finished query from scheduler and update min_query_id if needed, so that there aren't cancelled/finished queries in the scheduler.
     /// NOTE: call deleteQuery under the lock protection of MPPTaskManager
-    void deleteQuery(
-        const MPPQueryId & query_id,
-        MPPTaskManager & task_manager,
-        const bool is_cancelled,
-        Int64 gather_id);
+    void deleteQuery(const MPPQueryId & query_id, MPPTaskManager & task_manager, bool is_cancelled, Int64 gather_id);
 
     /// all scheduled tasks should finally call this function to release threads and schedule new tasks
-    void releaseThreadsThenSchedule(const int needed_threads, MPPTaskManager & task_manager);
+    void releaseThreadsThenSchedule(
+        const String & resource_group_name,
+        int needed_threads,
+        MPPTaskManager & task_manager);
 
 private:
     bool scheduleImp(
         const MPPQueryId & query_id,
         const MPPGatherTaskSetPtr & query_task_set,
         MPPTaskScheduleEntry & schedule_entry,
-        const bool isWaiting,
+        bool isWaiting,
         bool & has_error);
-    bool updateMinQueryId(const MPPQueryId & query_id, const bool retired, const String & msg);
-    void scheduleWaitingQueries(MPPTaskManager & task_manager);
-    bool isDisabled() { return thread_hard_limit == 0 && thread_soft_limit == 0; }
-    std::set<MPPQueryId> waiting_set;
-    std::set<MPPQueryId> active_set;
-    MPPQueryId min_query_id;
+    bool isDisabled() const { return thread_hard_limit == 0 && thread_soft_limit == 0; }
+
+    struct GroupEntry
+    {
+        explicit GroupEntry(const String & name)
+            : resource_group_name(name)
+            , min_query_id(MPPTaskId::Max_Query_Id)
+            , estimated_thread_usage(0)
+        {}
+
+        const String resource_group_name;
+        std::set<MPPQueryId> waiting_set;
+        std::set<MPPQueryId> active_set;
+        MPPQueryId min_query_id;
+        UInt64 estimated_thread_usage;
+
+        bool updateMinQueryId(const MPPQueryId & query_id, bool retired, const String & msg, LoggerPtr log);
+    };
+
+    void scheduleWaitingQueries(GroupEntry & entry, MPPTaskManager & task_manager, LoggerPtr log);
+    GroupEntry & mustGetGroupEntry(const String & resource_group_name);
+    GroupEntry & getOrCreateGroupEntry(const String & resource_group_name);
+
+    std::unordered_map<String, GroupEntry> scheduler_entries;
```

Comment:
Why not name it as `group_entries`? Since there is already a schedule_entry of type `MPPTaskScheduleEntry`

Reply:
done